src/main/scala/com/cognite/sdk/scala/common/GzipSttpBackend.scala

```diff
@@ -0,0 +1,109 @@
+package com.cognite.sdk.scala.common
+
+import java.io.ByteArrayOutputStream
+import java.nio.charset.Charset
+import java.util.zip.GZIPOutputStream
+
+import com.softwaremill.sttp._
+import org.apache.commons.io.IOUtils
+
+class GzipSttpBackend[R[_], S](delegate: SttpBackend[R, S], val minimumSize: Int = 1000)
+    extends SttpBackend[R, S] {
+  import GzipSttpBackend._
+
+  @SuppressWarnings(Array("org.wartremover.warts.Equals"))
+  implicit final class AnyOps[A](self: A) {
+    def strictEq(other: A): Boolean = self == other
+  }
+
+  private def isGzipped(keyAndValue: Tuple2[String, String]) =
+    keyAndValue._1.equalsIgnoreCase(HeaderNames.ContentEncoding) &&
+      keyAndValue._2.equalsIgnoreCase("gzip")
+
+  private def isGzippedContent(keyAndValue: Tuple2[String, String]) =
+    keyAndValue._1.equalsIgnoreCase(HeaderNames.ContentType) &&
+      keyAndValue._2.toLowerCase.contains("/gzip")
+
+  override def send[T](request: Request[T, S]): R[Response[T]] = {
+    val (newBody, newContentLength) = request.body match {
+      case body: BasicRequestBody
+          if !request.headers.exists(isGzipped)
+            && !request.headers.exists(isGzippedContent) =>
+        compressBody(body, minimumSize)
+      // TODO: Add support for streaming bodies.
+      //       Will likely require us to be more strict about S, for example
+      //       restricting it to fs2.Stream and using fs2.compress or fs2.compression.
+      //  case StreamBody(s) =>
+      // TODO: Add support for multipart bodies, though that seems difficult with sttp,
+      //       since the multipart encoding happens after send(?).
+      case _ => (request.body, None)
+    }
+    val newRequest = if (newBody.strictEq(request.body)) {
+      request
+    } else {
+      request
+        .copy(body = newBody)
+        .header(HeaderNames.ContentEncoding, "gzip")
+    }
+    val newRequestWithContentLength = newContentLength match {
+      case Some(contentLength) =>
+        newRequest
+          .copy(
+            headers = newRequest.headers.filterNot(_._1.equalsIgnoreCase(HeaderNames.ContentLength))
+          )
+          .header(HeaderNames.ContentLength, contentLength.toString)
```

Comment:
I think this won't replace the existing header, which was set in lines 46?

Reply:
ContentLength vs ContentEncoding